File: python/techai_webutils/src/techai_webutils/execution/job/job.py
"""Job[T]: a discoverer-backed, parallelised, optionally-retried unit of work.

The Python analog of Go's ``job.Job[T]`` — discover work items, fan each one out through
a processor with bounded concurrency, and optionally retry on failure or downgrade
failures to warnings. Retry is internal and async (mirroring Go's ``retryDo``) so this
package never depends on ``foundation`` and stays core-only.
"""

from __future__ import annotations

import asyncio
from dataclasses import replace
from typing import TYPE_CHECKING, Self

from techai_webutils.core.interfaces.execution import BatchResult, JobMeta, StepResult, StepStatus
from techai_webutils.execution.engine.fan_out import fan_out

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable

    from techai_webutils.core.interfaces.execution import Discoverer, ExecutionObserver

# Processor: transforms one discovered item into a StepResult; may raise to signal a
# retriable failure (the internal retry re-runs it; a final raise is isolated by fan_out).
type Processor[T] = Callable[[T], Awaitable[StepResult]]


class Job[T]:
    """Discovers items of type T, maps each through a processor, and fans them out.

    Construct via ``JobBuilder``. ``execute`` returns a single SKIP result when discovery
    is empty (mirroring Go ``Job.Execute``), otherwise the fanned-out BatchResult with
    failures optionally downgraded to warnings.
    """

    def __init__(
        self,
        name: str,
        discoverer: Discoverer[T],
        processor: Processor[T],
        *,
        group: str = "",
        concurrency: int = 0,
        retries: int = 0,
        retry_delay: float = 0.0,
        warn_only: bool = False,
        observer: ExecutionObserver | None = None,
    ) -> None:
        """Store the job configuration; prefer ``JobBuilder`` for construction."""
        # name labels the job in results/logs and names the empty-discovery SKIP.
        self._name = name
        # group buckets the job under a phase/category label in gate output + metrics.
        self._group = group
        # discoverer yields the work items to fan out.
        self._discoverer = discoverer
        # processor maps one item to a StepResult (may raise to signal a retriable failure).
        self._processor = processor
        # concurrency caps simultaneous items (<= 0 falls back to the engine default).
        self._concurrency = concurrency
        # retries is the number of extra attempts on a raised failure (0 = no retry).
        self._retries = retries
        # retry_delay is the pause (seconds) between retry attempts.
        self._retry_delay = retry_delay
        # warn_only downgrades FAIL results to WARN so the job is advisory/non-blocking.
        self._warn_only = warn_only
        # observer receives the fan-out lifecycle events (None = unobserved).
        self._observer = observer

    def meta(self) -> JobMeta:
        """Return the job's identity so a built Job satisfies ``AnyJob`` (mirrors Go ``Job.Meta``)."""
        return JobMeta(name=self._name, group=self._group)

    async def execute(self) -> BatchResult:
        """Discover items and fan them out; empty discovery yields a single SKIP.

        The empty-discovery SKIP is a job-level signal returned directly (not fanned out),
        so it deliberately bypasses the per-item observer — a SKIP marker is not a
        completed work item and must not increment the observer's completed/inflight
        series. Only a real fan-out (>= 1 item) drives the observer.
        """
        items = list(await self._discoverer.discover())
        if not items:
            return BatchResult(
                results=(StepResult(name=self._name, group=self._group, status=StepStatus.SKIP),),
            )

        batch = await fan_out(
            items,
            self._concurrency,
            self._processor_with_retry(),
            self._observer,
            name=self._name,
        )
        if self._warn_only:
            batch = _downgrade_failures(batch)
        return batch

    def _processor_with_retry(self) -> Processor[T]:
        """Return the processor, wrapped with an internal async retry when retries > 0.

        The retry fires only on a *raised* exception (mirroring Go's retry-on-error); a
        processor that instead *returns* ``StepResult(FAIL)`` — the value fan_out isolates
        raised errors into — is treated as a terminal outcome and is NOT retried.
        """
        if self._retries <= 0:
            return self._processor

        processor = self._processor
        retries = self._retries
        delay = self._retry_delay

        async def retried(item: T) -> StepResult:
            """Re-run processor up to ``retries`` extra times on failure (mirrors Go retryDo)."""
            attempt = 0
            while True:
                try:
                    return await processor(item)
                except asyncio.CancelledError:
                    raise  # never retry cancellation
                except Exception:
                    if attempt >= retries:
                        raise
                    attempt += 1
                    if delay > 0:
                        await asyncio.sleep(delay)

        return retried


def _downgrade_failures(batch: BatchResult) -> BatchResult:
    """Downgrade every FAIL result to WARN (mirrors Go WarnOnly)."""
    return BatchResult(
        results=tuple(
            replace(r, status=StepStatus.WARN) if r.status is StepStatus.FAIL else r for r in batch.results
        ),
    )


class JobBuilder[T]:
    """Fluent builder for a ``Job[T]`` (mirrors Go's ``job.Builder[T]``)."""

    def __init__(self, name: str) -> None:
        """Start a builder for a job with the given name."""
        self._name = name
        self._group = ""
        self._discoverer: Discoverer[T] | None = None
        self._processor: Processor[T] | None = None
        self._concurrency = 0
        self._retries = 0
        self._retry_delay = 0.0
        self._warn_only = False
        self._observer: ExecutionObserver | None = None

    def grouped(self, group: str) -> Self:
        """Set the job's group label (phase/category) carried in its JobMeta."""
        self._group = group
        return self

    def with_discoverer(self, discoverer: Discoverer[T]) -> Self:
        """Set the discoverer that yields work items."""
        self._discoverer = discoverer
        return self

    def with_processor(self, processor: Processor[T]) -> Self:
        """Set the async processor applied to each discovered item."""
        self._processor = processor
        return self

    def with_concurrency(self, n: int) -> Self:
        """Set the maximum number of items processed concurrently."""
        self._concurrency = n
        return self

    def with_retry(self, retries: int, delay: float = 0.0) -> Self:
        """Set the number of extra retry attempts on failure and the pause between them."""
        self._retries = retries
        self._retry_delay = delay
        return self

    def warn_only(self, *, enabled: bool = True) -> Self:
        """Mark the job non-blocking: FAIL results are downgraded to WARN."""
        self._warn_only = enabled
        return self

    def with_observer(self, observer: ExecutionObserver) -> Self:
        """Set the execution observer that receives lifecycle events."""
        self._observer = observer
        return self

    def build(self) -> Job[T]:
        """Construct the Job[T]; raises ValueError if a discoverer or processor is missing."""
        if self._discoverer is None or self._processor is None:
            msg = "JobBuilder requires both a discoverer and a processor"
            raise ValueError(msg)
        return Job(
            self._name,
            self._discoverer,
            self._processor,
            group=self._group,
            concurrency=self._concurrency,
            retries=self._retries,
            retry_delay=self._retry_delay,
            warn_only=self._warn_only,
            observer=self._observer,
        )

File: go/execution/job/job.go
// Package job provides the Job[T] type: a discoverer-backed, parallelised,
// optionally-retried unit of work implementing interfaces.AnyJob.
package job

import (
	"context"
	"fmt"
	"path/filepath"
	"time"

	coreerr "github.com/gt-tech-ai/knowledge-engine/go/core/errors"
	"github.com/gt-tech-ai/knowledge-engine/go/core/interfaces"
	"github.com/gt-tech-ai/knowledge-engine/go/core/types"
	"github.com/gt-tech-ai/knowledge-engine/go/execution/engine"
)

// JobConfig holds execution parameters applied to every step in the job.
type JobConfig struct {
	// Concurrency is the max number of parallel workers; 0 means runtime.NumCPU.
	Concurrency int

	// Retry is the max number of retry attempts after a step failure (0 = no retry).
	Retry int

	// RetryDelay is the pause between retry attempts.
	RetryDelay time.Duration

	// WarnOnly downgrades Fail results to Warn so the job never blocks a gate.
	WarnOnly bool
}

// Option configures a JobConfig.
type Option func(*JobConfig)

// WithConcurrency sets the maximum number of parallel workers.
func WithConcurrency(n int) Option { return func(c *JobConfig) { c.Concurrency = n } }

// WithRetry sets the maximum number of retry attempts on step failure.
func WithRetry(n int) Option { return func(c *JobConfig) { c.Retry = n } }

// WithRetryDelay sets the pause between retry attempts.
func WithRetryDelay(
	d time.Duration,
) Option {
	return func(c *JobConfig) { c.RetryDelay = d }
}

// WithWarnOnly marks the job non-blocking: Fail results are downgraded to Warn.
func WithWarnOnly() Option { return func(c *JobConfig) { c.WarnOnly = true } }

// Job[T] discovers work items of type T, maps each to a WorkUnit, and
// fans them out in parallel. It implements interfaces.AnyJob.
// Field order minimises GC pointer-scan range: interfaces, func, strings, value types.
type Job[T any] struct {
	// discoverer yields the work items of type T at execution time.
	discoverer interfaces.Discoverer[T]

	// tool is the optional required tool; when set its binary is checked via
	// RequireTool and a missing binary turns the whole job into a skip.
	tool interfaces.ToolSpec

	// mapper converts each discovered item into the WorkUnit that is run.
	mapper func(root string, item T) engine.WorkUnit

	// name identifies the job (and single-unit steps) in StepResults.
	name string

	// group categorizes the job for display grouping.
	group string

	// config holds the concurrency, retry, and warn-only execution parameters.
	config JobConfig
}

// Compile-time assertion that *Job[string] satisfies interfaces.AnyJob.
var _ interfaces.AnyJob = (*Job[string])(nil)

// Builder[T] assembles a Job[T] via a fluent API.
// Field order minimises GC pointer-scan range: interfaces, func, strings, slices.
type Builder[T any] struct {
	// discoverer is the item source set via WithDiscoverer.
	discoverer interfaces.Discoverer[T]

	// tool is the optional required tool set via WithTool.
	tool interfaces.ToolSpec

	// mapper is the item-to-WorkUnit function set via WithMapper.
	mapper func(root string, item T) engine.WorkUnit

	// name is the job name carried into the built Job.
	name string

	// group is the display group carried into the built Job.
	group string

	// opts accumulate the execution options applied to JobConfig on Build.
	opts []Option
}

// New returns a Builder[T] for a job with the given name and group.
func New[T any](name, group string) *Builder[T] {
	return &Builder[T]{name: name, group: group}
}

// WithTool sets the ToolSpec checked by RequireTool before execution.
func (b *Builder[T]) WithTool(t interfaces.ToolSpec) *Builder[T] {
	b.tool = t
	return b
}

// WithDiscoverer sets the Discoverer that yields work items.
func (b *Builder[T]) WithDiscoverer(d interfaces.Discoverer[T]) *Builder[T] {
	b.discoverer = d
	return b
}

// WithMapper sets the function that converts (root, item) into a WorkUnit.
func (b *Builder[T]) WithMapper(
	fn func(root string, item T) engine.WorkUnit,
) *Builder[T] {
	b.mapper = fn
	return b
}

// WithOptions appends execution options.
func (b *Builder[T]) WithOptions(opts ...Option) *Builder[T] {
	b.opts = append(b.opts, opts...)
	return b
}

// Build constructs and returns the Job[T].
func (b *Builder[T]) Build() interfaces.AnyJob {
	cfg := JobConfig{}
	for _, o := range b.opts {
		o(&cfg)
	}
	return &Job[T]{
		discoverer: b.discoverer,
		mapper:     b.mapper,
		name:       b.name,
		group:      b.group,
		tool:       b.tool,
		config:     cfg,
	}
}

// Meta implements interfaces.AnyJob.
func (j *Job[T]) Meta() types.JobMeta {
	toolName := ""
	if j.tool != nil {
		toolName = j.tool.Binary()
	}
	return types.JobMeta{Name: j.name, Group: j.group, Tool: toolName}
}

// Execute implements interfaces.AnyJob.
// It checks the required tool (skip if missing), discovers items (skip if none),
// maps each to a WorkUnit, fans them out via FanOut, then optionally downgrades
// failures to warnings when WarnOnly is set.
func (j *Job[T]) Execute(
	ctx context.Context,
	runner interfaces.CommandRunner,
	root string,
) (types.StepResults, error) {
	if j.tool != nil && j.tool.Binary() != "" {
		if err := runner.RequireTool(j.tool.Binary(), j.tool.InstallHint()); err != nil {
			// Missing tool is a skip, not a fatal error - the job is optional.
			return types.StepResults{ //nolint:nilerr // intentional: missing tool => skip, not error
				{Name: j.name, Status: types.StatusSkip, Error: err.Error()},
			}, nil
		}
	}

	items, err := j.discoverer.Discover(ctx, root)
	if err != nil {
		return nil, err
	}
	if len(items) == 0 {
		return types.StepResults{{Name: j.name, Status: types.StatusSkip}}, nil
	}

	units := make([]engine.WorkUnit, len(items))
	for i, item := range items {
		units[i] = j.mapper(root, item)
	}

	results := engine.FanOut(
		ctx,
		units,
		j.config.Concurrency,
		func(ctx context.Context, wu engine.WorkUnit) types.StepResult {
			var runErr error
			if j.config.Retry > 0 {
				runErr = retryDo(ctx, func() error {
					return runner.Run(ctx, wu.Dir, wu.Exe, wu.Args...)
				}, j.config.Retry, j.config.RetryDelay)
			} else {
				runErr = runner.Run(ctx, wu.Dir, wu.Exe, wu.Args...)
			}
			r := types.StepResult{Name: wu.Name, Status: types.StatusPass}
			if runErr != nil {
				r.Status = types.StatusFail
				r.Error = runErr.Error()
			}
			return r
		},
	)

	if j.config.WarnOnly {
		for i := range results {
			if results[i].Status == types.StatusFail {
				results[i].Status = types.StatusWarn
			}
		}
	}

	return results, nil
}

// Single creates a Job[string] that runs exactly one command.
// tool.Cmd overrides the executable prefix when set (e.g. ["uv", "run", "ruff"]).
func Single(
	name, group string,
	tool interfaces.ToolSpec,
	argsFn func(root string) []string,
	dirFn func(root string) string,
	opts ...Option,
) interfaces.AnyJob {
	disc := interfaces.DiscovererFunc[string](
		func(_ context.Context, _ string) ([]string, error) {
			return []string{name}, nil
		},
	)

	mapFn := func(root, _ string) engine.WorkUnit {
		dir := dirFn(root)
		if !filepath.IsAbs(dir) {
			dir = filepath.Join(root, dir)
		}
		exe, args := tool.Cmd(argsFn(root)...)
		return engine.WorkUnit{Name: name, Dir: dir, Exe: exe, Args: args}
	}
	return New[string](name, group).
		WithTool(tool).
		WithDiscoverer(disc).
		WithMapper(mapFn).
		WithOptions(opts...).
		Build()
}

// retryDo calls fn up to maxRetries additional times after a failure.
func retryDo(
	ctx context.Context,
	fn func() error,
	maxRetries int,
	delay time.Duration,
) error {
	var last error
	for attempt := 0; attempt <= maxRetries; attempt++ {
		if attempt > 0 && delay > 0 {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(delay):
			}
		}
		last = fn()
		if last == nil {
			return nil
		}
	}
	return coreerr.Wrap(
		last,
		coreerr.CodeInternal,
		fmt.Sprintf("failed after %d attempt(s)", maxRetries+1),
	)
}
